### ISSUE TITLE
Test Vectors not matching Expected Result due to ambiguity of uncertainty points

### ISSUE THREAD

**Issue by alexcpn** [Wireless-Innovation-Forum/6-GHz-AFC#68]
As per R2-AIP-22 of WINNF-TS-1014-V1.5, uncertainty points about an Access Point should be sampled at intervals no greater than 1 arc second

"The evaluation points within this uncertainty volume shall be determined based on a
grid superimposed on the Standard Power Device’s horizontal area of location uncertainty as reported in the request, along with the reported height HSPD (AGL) and associated vertical uncertainty HSPD, sampled at intervals no greater than 1 arc second
of latitude and longitude and no greater than 5 meters of elevation."

But in doing so, there is no guarantee that the lowest PSD point will be found. ( Just lowest propagation loss may not lead to the lowest PSD point as RPE could change based on angles, so to be accurate PSD need to be calculated for all uncertainty points)

To illustrate this take the case of  FSP.13

##  FSP.13 Test Vector - Distance between Uncertainty points at one Arc Second

Black dot is receiver. Red dot is the point where loss is minimum, this is not there in current yellow dot uncertainty
Cross is the center, original AP co-ordinates

With 1 arc second uncertainty

![image](https://github.com/user-attachments/assets/74acba33-5891-4acf-b29e-5c568c397f0d)

```
[INFO] Going to process 528/3262 UUID 1053054,KMJ53,CF=6063.8 Mhz,1053054KMJ53P[50],T[L=1,A=50],R[L=28,A=2],CF=6063.8,fr_id=1,Dist=11.28 m
[INFO] Permissible PSD =-12.443606210807118 dBm/MHz [PSD Receiver=-12.443606210807118 PSD Diversity=None PSD Passive=None]
[INFO] Loss Receivers=76.35639378919288 Mode=-fspl at AP Loc=(33.76974577295118, -118.37634107953944) dist=25.881307425798973 GRx_Effective=-3.700000000000003 gamma=39.18182498297076 accesspoint_ht_m=5.0
[INFO] receiver_clutterloss=0 ap_nlcd_code=0  ap_clutterloss=0 freq=(6048.8, 6078.8)
[INFO] uuid: 1053054
```
Plotted on QGIS and visually on scale we can see that points are 1 arc second (approx 30m) away

![image](https://github.com/user-attachments/assets/3192e628-4839-4090-b9dc-a71da81d6f0f)

With this, the lowest PSD is -12.44 and it is greater than 8.36 dB from upper bound

```
[ERROR] Calculated PSD -12.443606210807118 Greater-Than 8.36 dB from upper bound [-20.8] for  (6048, 6079)
```

## Reducing distance between Uncertainty points to 1/4th of an Arc Second

Reducing distance between Uncertainty points to 1/4th of an Arc Second is about 30/4 which is 7 meters

![image](https://github.com/user-attachments/assets/e62db655-3c1d-40a4-8f13-b9e23d9af5fd)

Which gives a much closer result 

```
[INFO] Going to process 6/27 UUID 1053054,KMJ53,CF=6063.8 Mhz,1053054KMJ53P[50],T[L=1,A=50],R[L=28,A=2],CF=6063.8,fr_id=1,Dist=11.28 m
[INFO] Permissible PSD =-19.06281661641762 dBm/MHz [PSD Receiver=-19.06281661641762 PSD Diversity=None PSD Passive=None]
[INFO] Loss Receivers=74.79718338358238 Mode=-fspl at AP Loc=(33.769676328506726, -118.37627163509498) dist=15.752761456136097 GRx_Effective=1.3599999999999968 gamma=43.996094068222106 accesspoint_ht_m=5.0
[INFO] receiver_clutterloss=0 ap_nlcd_code=0  ap_clutterloss=0 freq=(6048.8, 6078.8)
```

Though slightly higher

```
[ERROR] Calculated PSD -19.06281661641762 Greater-Than 1.74 dB from upper bound [-20.8] for  (6048, 6079)
```

## Summary

To get results as per the test vector output, for some test vectors, we need to arrive at the same uncertainty point that the test authors have somehow derived.

Following 1 arc second specification that is 30 meters equidistant does not guarantee that we get the same AP as the test vector. Making it dense, increases the number of computations severalfold and increases the probability of getting closer to what the test authors have derived.

**Comment by alexcpn** [comment]
I changed the code as per AFC SUT Test Plan v1.5 ,  Appendix A.1.5
[AFC System (SUT) Compliance Test Plan v1.5.pdf](https://github.com/user-attachments/files/18078669/AFC.System.SUT.Compliance.Test.Plan.v1.5.pdf)

_

> Spectrum availability shall be evaluated across the entire volume of reported AFC Device uncertainty at
> resolutions of 1 arc second horizontally and 5 m vertically. The evaluation points shall be selected horizontally as
> the center points of 1x1 arc second cells in which integer lines of latitude and longitude fall at the edge of a cell
> (pixel) along 1x1-degree tile boundaries and the 1 arc second pixel centers of longitude are spaced at integer arc
> second values, and for every cell with which the reported AFC Device uncertainty overlaps
_

The value is still off. Plotting this on QGIS, the  yellow and green points are FSP.13 uncertainty as per the  new code. The bright red is the point which has the least loss and it is not in the new uncertainty points as well. 
The lowest PSD is the  green uncertainty point which gives the lowest PSD of  ` -15.13374865996191 `

```
[ERROR] Calculated PSD -15.13374865996191 Greater-Than 5.67 dB from upper bound [-20.8] for  (6048, 6079)
```

```
 Going to process 528/3262 UUID 1053054,KMJ53,CF=6063.8 Mhz,1053054KMJ53P[50],T[L=1,A=50],R[L=28,A=2],CF=6063.8,fr_id=1,Dist=11.28 m
[INFO] Permissible PSD =-15.13374865996191 dBm/MHz [PSD Receiver=-15.13374865996191 PSD Diversity=None PSD Passive=None]
[INFO] Loss Receivers=73.3402513400381 Mode=fspl-e at AP Loc=(33.76958333333334, -118.37625) dist=10.743871563938074 GRx_Effective=-4.025999999999995 gamma=64.57759107189443 accesspoint_ht_m=5.0
[INFO] 
```
![image](https://github.com/user-attachments/assets/2f665c40-b368-4868-a122-826dac27539a)

**Comment by alexcpn** [comment]
By running with the updated uncertainty points I am hitting the same problem with FSP.62

The red and blue points are the uncertainty points in the centre of 1 arc second grids. The black point is the receiver. 
the blue point is the one with the least PSD among the calculated  uncertainty points. However there is a point (green dot) which has the lowest loss and which does not fall in the calculated uncertainty

![image](https://github.com/user-attachments/assets/9f7d6892-cd4f-4723-a437-11d8b7d81f88)


### Blue point: 'PSD =-14.652170368651923'

```
 Going to process 1104/2905 UUID 1051886,WQU78,CF=6226.89 Mhz,1051886WQU78P[1],T[L=1,A=1],R[L=3,A=1],CF=6226.89,fr_id=1,Dist=25.68 m
[INFO] Permissible PSD =-14.652170368651923 dBm/MHz [PSD Receiver=-14.652170368651923 PSD Diversity=None PSD Passive=67.53624686006555]
[INFO] Loss Receivers=71.04782963134808 Mode=fspl-e at AP Loc=(38.82319444444445, -120.68458333333334) dist=13.36732981521102 GRx_Effective=-6.799999999999997 gamma=94.19871597938736 accesspoint_ht_m=5.0
[INFO] Loss Passive Receivers=124.14251137621845 Mode=itm-fspl  ap_clutterloss= -0.3197464167111226 at AP Loc=(38.82260924141299, -120.68442934311247) 
[INFO] Permissible Passive PSD =67.53624686006555 GVRxDisc=-47.68788645194361 GVRxEffective=-9.025595448614617 gamma=86.91331745988185
[INFO] receiver_clutterloss=0 ap_nlcd_code=0  ap_clutterloss=0 freq=(6211.89, 6241.89)
```

### Green point: `PSD =-26.169834900591113 `

```
INFO] Going to process 1104/2905 UUID 1051886,WQU78,CF=6226.89 Mhz,1051886WQU78P[1],T[L=1,A=1],R[L=3,A=1],CF=6226.89,fr_id=1,Dist=25.68 m
[INFO] Receiver in AP uncertanity: Calculating new uncertanity points len=1098
[INFO] Ks =0.005541538389603947 1/Ks=180.45530495214524 alpha_n=-47.22559544861462 db area=14.700000000000001  dPathSegLength=6226.178359923841 
[INFO] G_pathseg capped at -47.22559544861462
[INFO] GVRxEffective=-9.025595448614617 
[INFO] Permissible PSD =-26.169834900591113 dBm/MHz [PSD Receiver=-26.169834900591113 PSD Diversity=None PSD Passive=67.53624686006555]
[INFO] Loss Receivers=59.53016509940889 Mode=fspl-e at AP Loc=(38.82328953051957, -120.68449242924363) dist=2.1866830304162677 GRx_Effective=-6.799999999999997 gamma=99.0837744160475 accesspoint_ht_m=5.0
[INFO] Loss Passive Receivers=124.14251137621845 Mode=itm-fspl  ap_clutterloss= -0.3197464167111226 at AP Loc=(38.82260924141299, -120.68442934311247) 
[INFO] Permissible Passive PSD =67.53624686006555 GVRxDisc=-47.68788645194361 GVRxEffective=-9.025595448614617 gamma=86.91331745988185
[INFO] receiver_clutterloss=0 ap_nlcd_code=0  ap_clutterloss=0 freq=(6211.89, 6241.89)
[INFO] uuid: 1051886
```
Shape files for plotting in say QGIS 
[fsp62_shp_files.zip](https://github.com/user-attachments/files/18106661/fsp62_shp_files.zip)

**Comment by AEgbert** [comment]
Do the changes to the grid from the SUT Test Plan A.1.5 help with issue #66? Looking at that issue led me to suggest A.1.5, since the FS receiver was about 18km away. I didn't notice that in this issue (#68) the horizontal uncertainty region of the inquiry overlaps the incumbent location. For instances like these (FS receiver within the Standard Power Device's horizontal uncertainty footprint), there's additional guidance in R2-AIP-40 of TS-1014. The SUT Test Plan A.1.15 echoes the same process.

**Comment by alexcpn** [comment]
@AEgbert the R2-AIP-40 does not apply in this case as I mentioned 

>Note that r_height_to_center_raat: 7.9 is greater than max AP height 3+2=5m and hence R2-AIP-40 a does not apply

>>if the FS Rx height is inside the Standard Power Device’s vertical uncertainty, AFC System reports an allowable PSD2 level of -40 dBm/MHz for the frequencies co-channel with that FS Rx

Also with the updated code I ran for #66  FSP.37 and FSP.44 test cases and it is not helping

```
[INFO] Going to process 2901/3523 UUID 1053054,KMJ53,CF=6715.0 Mhz,1053054KMJ53P[43],T[L=1,A=43],R[L=26,A=1],CF=6715.0,fr_id=1,Dist=18107.25 m
[INFO] Permissible PSD =19.664086556122555 dBm/MHz [PSD Receiver=19.664086556122555 PSD Diversity=None PSD Passive=None]
[INFO] Loss Receivers=134.17188976686444 Mode=itm-fspl at AP Loc=(34.05236111111111, -118.17402777777778) dist=18174.20444400239 GRx_Effective=30.366059190609498 gamma=1.094553369953994 accesspoint_ht_m=11.0
[INFO] receiver_clutterloss=0 ap_nlcd_code=23  ap_clutterloss=28.358255979867597 freq=(6710.0, 6720.0)
```

**Comment by AEgbert** [comment]
R2-AIP-40-a applies when the device is in both the vertical and horizontal uncertainty. However, R2-AIP-40-b applies if it is within the horizontal uncertainty but not the vertical uncertainty. If I understand your map correctly, the receiver is within the standard power device horizontal uncertainty.

**Comment by alexcpn** [comment]
>Compute the minimum angle-off-boresight from that FS Rx to the Standard Power Device’s uncertainty volume, and use this angle to compute FS Rx antenna gain towards the Standard Power Device for all evaluation points.

Just to be clear; what does `Compute the minimum angle-off-boresight from that FS Rx to the Standard Power Device’s uncertainty volume,`  mean ? Does the uncertainty volume here means evaluation points

That is first calculate from the receiver to all uncertainty points the minimum angle-off-boresight; and then using that angle calculate gain and PSD for all evaluation points ?

What I have done is taken the min PSD , not the initial minimum angle-off-broresight. I will check with that

**Comment by alexcpn** [comment]
@AEgbert  I added this handling of calculating the minimum off-boresight angle (gamma ) for receivers which are in the uncertainty horizontally of the AP and keeping that constant and using that to calculate the PSD for all uncertainty points; instead of calculating the PSD for each uncertainty location and taking the lowest PSD.

However, the results are worse.  The result is for FSP.13 for receiver UUDI 1053054 Callsign KMJ53 and  freq=(6048.8, 6078.8)

The result for this receiver and frequency band
[1053054_KMJ53_-10.081748659961903.xls](https://github.com/user-attachments/files/18135381/1053054_KMJ53_-10.081748659961903.xls)

```
[INFO] Expected Frequency
[INFO] 6048-6079,upperBound:-20.8 nominalValue:-22.8
[INFO] Calculated Frequency
[INFO] (6048, 6079)  psd: -10.081748659961903
[INFO] --------------------------------------------------------------------------------
[ERROR] Calculated PSD -10.081748659961903 Greater-Than 10.72 dB from upper bound [-20.8] for  (6048, 6079)
```

```
[INFO] ********************************************************************************
[INFO] Going to process 6/27 UUID 1053054,KMJ53,CF=6063.8 Mhz,1053054KMJ53P[50],T[L=1,A=50],R[L=28,A=2],CF=6063.8,fr_id=1,Dist=11.28 m
[INFO] Receiver in AP horizontal uncertanity minimum angle off-boresight is 23.880839914457287 dregrees
[INFO] Permissible PSD =-10.081748659961903 dBm/MHz [PSD Receiver=-10.081748659961903 PSD Diversity=None PSD Passive=None]
[INFO] Loss Receivers=73.3402513400381 Mode=fspl-e at AP Loc=(33.76958333333334, -118.37625) dist=10.743871563938074 GRx_Effective=-9.078000000000003 gamma=23.880839914457287 accesspoint_ht_m=5.0
[INFO] receiver_clutterloss=0 ap_nlcd_code=0  ap_clutterloss=0 freq=(6048.8, 6078.8)
[INFO] uuid: 1053054
[INFO] call_sign: KMJ53
[INFO] frequency_assigned: 6063.8
[INFO] frequency_upper_band: None
[INFO] transmitter_make: Aviat Networks, Inc.
[INFO] transsmitter_model: I600V3EHPL6-30M 256Q 179M
[INFO] is_odu: False
[INFO] receiver_loc: (33.769555555555556, -118.37636111111111)
[INFO] t_loc_num: 1
[INFO] t_path_num: 50
[INFO] t_ant_num: 50
[INFO] fr_id: 1
[INFO] receiver_to_ap_dist_m: 11.29656922500915
[INFO] r_height_to_center_raat: 19.8
[INFO] r_diversity_gain: None
[INFO] r_azimuth: 0.0
[INFO] r_antenna_make: Commscope
[INFO] r_antenna_model: UHX8-59H RF
[INFO] r_loc_num: 28
[INFO] r_path_num: 50
[INFO] r_ant_num: 2
[INFO] r_polarization_code: None
[INFO] line_loss: None
[INFO] r_diameter_ft: 8.005249599999999
[INFO] r_diameter_m: 2.44
[INFO] antenna_type_code: R
[INFO] ant_category: HP
[INFO] r_diversity_height: 8.8
[INFO] transmitter_loc: (34.05333333333333, -118.16977777777778)
[INFO] dFSTx_FSRx_m: 36866.26229526948
[INFO] t_height_to_center_raat: 35.0
[INFO] t_gain: 41.3
[INFO] t_diversity_gain: None
[INFO] t_azimuth: 211.3
[INFO] t_antenna_make: Commscope
[INFO] t_antenna_model: UHX8-59H RF
[INFO] t_polarization_code: V    
[INFO] uls_bandwidth: 30.0
[INFO] bandwidth: 30.0
[INFO] center_freq: 6063.8
[INFO] frequency_range: (6048.8, 6078.8)
[INFO] r_gain: 41.3
[INFO] accesspoint_loc: (33.769641, -118.376295)
[INFO] accesspoint_ht_m: 3.0
[INFO] receiver_elev: 368.8902124017477
[INFO] transmitter_elev: 176.6184539794922
[INFO] ********************************************************************************
```

**Comment by alexcpn** [comment]
Getting the Uncertainty correctly is very important to get the valid results. Unfortunately it is not specified in WINNF TS-1014 how this has to be handled which I feel is very vital

The comments  from https://github.com/Wireless-Innovation-Forum/6-GHz-AFC/issues/66#issuecomment-2553918349 provide vital clues

*Also, your sample point grid looks like it got rotated. Note that we had to figure out during our process to align with the test vectors that all evaluation points need to be on the half second intervals, of the lat/lon grid, such as 00.5", 01.5", 02.5", skipping the integer values. This places the selection of the nlcd values in the dead center of the cell. We also do not include as many points around the edge of the ellipse. Instead, if any corner of a 1" cell overlaps the ellipse, that center is included, so you will end up using points a little bit outside of the ellipse in your calculations. This algorithm is also NOT in TS-1014.*

One thing still not clear is that when the orientation is > 0, should the grid boxes be rotated and the center taken ; or not

Example rotated grid and centered (is this the right one; I am assuming so)

![Image](https://github.com/user-attachments/assets/1af3e6f9-1e4a-4062-9c43-be06afa571ea)

than not rotating the grid like below and uncertainty points in the center of the gird

![Image](https://github.com/user-attachments/assets/6faf963a-0693-4015-8c32-9b941565fff8)
